File: CoreLibrary/Package/GUI_Components/Controls/_Validation/InputtedValueValidation.ts
/* eslint-disable max-classes-per-file --
* This limitation is unsolicited for the namespaced classes, however, there is no ESLint option allowing this case. */
import { InvalidParameterValueError, isBoolean, Logger } from "@yamato-daiwa/es-extensions";

import inputtedValueValidationLocalization__english from "./InputtedValueValidationLocalization.english";


abstract class InputtedValueValidation<NonEmptyValueType, EmptyValueType = NonEmptyValueType> {

  /* ━━━ Static Fields ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public static localization: InputtedValueValidation.Localization = inputtedValueValidationLocalization__english;


  /* ━━━ Instance Fields ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  /* ┅┅┅ Instance ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅ */
  public readonly isValueOfSupportedType:
      InputtedValueValidation.SupportedValueChecker<NonEmptyValueType, EmptyValueType>;

  public readonly hasValueBeenOmitted:
      InputtedValueValidation.OmittedValueChecker<NonEmptyValueType, EmptyValueType>;

  public readonly isInputRequired: InputtedValueValidation.InputRequirementChecker;

  protected readonly requiredInputIsMissingValidationErrorMessage: string;
  protected readonly staticRules: ReadonlyArray<InputtedValueValidation.Rule<NonEmptyValueType>>;
  protected readonly contextDependentRules: ReadonlyArray<InputtedValueValidation.Rule<NonEmptyValueType>>;
  protected readonly asynchronousRules: ReadonlyArray<InputtedValueValidation.AsynchronousRule<NonEmptyValueType>>;


  /* ━━━ Constructor━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  protected constructor(
    {
      isValueOfSupportedType,
      hasValueBeenOmitted,
      isInputRequired,
      requiredInputIsMissingValidationErrorMessage,
      localization,
      staticRules,
      contextDependentRules,
      asynchronousRules
    }: InputtedValueValidation.ConstructorCompoundParameter<NonEmptyValueType, EmptyValueType>
  ) {

    this.isValueOfSupportedType = isValueOfSupportedType;
    this.hasValueBeenOmitted = hasValueBeenOmitted;
    this.isInputRequired = isBoolean(isInputRequired) ? (): boolean => isInputRequired : isInputRequired;

    this.requiredInputIsMissingValidationErrorMessage =
        requiredInputIsMissingValidationErrorMessage ??
        localization?.requiredInputIsMissingValidationErrorMessage ??
        InputtedValueValidation.localization.requiredInputIsMissingValidationErrorMessage;

    this.staticRules = staticRules ?? [];
    this.contextDependentRules = contextDependentRules ?? [];
    this.asynchronousRules = asynchronousRules ?? [];

  }


  public validate(
    rawValue: unknown,
    {
      mustPostponeAsynchronousValidation = false,
      asynchronousChecksCallback,
      messagesOfExternallyDetectedValidationErrors = []
    }: Readonly<{
      mustPostponeAsynchronousValidation?: boolean;
      asynchronousChecksCallback?: InputtedValueValidation.AsynchronousChecks.Callback;
      messagesOfExternallyDetectedValidationErrors?: ReadonlyArray<string>;
    }> = {}
  ): InputtedValueValidation.Result {

    if (!this.isValueOfSupportedType(rawValue)) {
      Logger.throwErrorWithFormattedMessage({
        errorInstance: new InvalidParameterValueError({
          parameterNumber: 1,
          parameterName: "rawValue",
          messageSpecificPart: "The type of `rawValue` is incompatible with specified validators."
        }),
        title: InvalidParameterValueError.localization.defaultTitle,
        occurrenceLocation: "inputtedValueValidation.validate(rawValue, options)",
        additionalData: {
          rawValue,
          nativeTypeOfRawValue: typeof rawValue
        }
      });
    }


    if (this.hasValueBeenOmitted(rawValue)) {

      return this.isInputRequired() ?
          {
            isValid: false,
            errorsMessages: [ this.requiredInputIsMissingValidationErrorMessage ]
          } :
          { isValid: true };

    }


    const validationErrorsMessages: Array<string> = [ ...messagesOfExternallyDetectedValidationErrors ];

    for (const staticValidationRule of this.staticRules) {

      const checkingResult: InputtedValueValidation.Rule.CheckingResult = staticValidationRule.check(rawValue);

      if (!checkingResult.isValid) {

        validationErrorsMessages.push(checkingResult.errorMessage);

        if (staticValidationRule.mustFinishValidationIfValueIsInvalid) {
          break;
        }

      }

    }

    if (validationErrorsMessages.length > 0) {
      return {
        isValid: false,
        errorsMessages: validationErrorsMessages
      };
    }


    for (const contextDependentValidationRule of this.contextDependentRules) {

      const checkingResult: InputtedValueValidation.Rule.CheckingResult = contextDependentValidationRule.check(rawValue);

      if (!checkingResult.isValid) {

        validationErrorsMessages.push(checkingResult.errorMessage);

        if (contextDependentValidationRule.mustFinishValidationIfValueIsInvalid) {
          break;
        }

      }

    }

    if (validationErrorsMessages.length > 0) {
      return {
        isValid: false,
        errorsMessages: validationErrorsMessages
      };
    }


    const validationResult: InputtedValueValidation.Result = { isValid: true };

    if (!mustPostponeAsynchronousValidation) {
      this.executeAsynchronousChecksIfAny(rawValue, validationResult, asynchronousChecksCallback);
    }


    return validationResult;

  }

  public executeAsynchronousChecksIfAny(
    rawValue: NonEmptyValueType,
    currentValidationResult: InputtedValueValidation.Result,
    asynchronousChecksCallback?: InputtedValueValidation.AsynchronousChecks.Callback
  ): void {

    if (this.asynchronousRules.length === 0) {
      return;
    }


    const asynchronousChecks: InputtedValueValidation.AsynchronousChecks = this.asynchronousRules.reduce(
      (
        accumulatingValue: InputtedValueValidation.AsynchronousChecks,
        asynchronousValidationRule: InputtedValueValidation.AsynchronousRule<NonEmptyValueType>
      ): InputtedValueValidation.AsynchronousChecks => {

        accumulatingValue[asynchronousValidationRule.ID] = {
          message: asynchronousValidationRule.messages.checkingInProgress,
          isPending: true,
          hasInvalidValueBeenConfirmed: false,
          hasValidValueBeenConfirmed: false,
          hasErrorOccurred: false
        };

        return accumulatingValue;

      },
      {}
    );

    asynchronousChecksCallback?.(
      new InputtedValueValidation.AsynchronousChecks.Status(asynchronousChecks),
      currentValidationResult
    );

    for (const validationRule of this.asynchronousRules) {

      validationRule.

          check(rawValue).

          then((checkingResult: InputtedValueValidation.AsynchronousRule.CheckingResult): void => {

						asynchronousChecks[validationRule.ID] = {
							isPending: false,
							hasValidValueBeenConfirmed: checkingResult.isValid,
							hasInvalidValueBeenConfirmed: !checkingResult.isValid,
							hasErrorOccurred: false,
							message: checkingResult.isValid ?
									validationRule.messages.validValueHasBeenConfirmed :
									checkingResult.errorMessage ?? validationRule.messages.invalidValueHasBeenConfirmed
						};

            const asynchronousCheckStatus: InputtedValueValidation.AsynchronousChecks.Status =
                new InputtedValueValidation.AsynchronousChecks.Status(asynchronousChecks);

            const errorsMessages: ReadonlyArray<string> = [
              ...currentValidationResult.isValid ? [] : currentValidationResult.errorsMessages,
              ...asynchronousCheckStatus.errorsMessages
            ];

						asynchronousChecksCallback?.(
              asynchronousCheckStatus,
              {
                isValid: errorsMessages.length === 0,
                errorsMessages
              }
            );

					}).

          catch((error: unknown): void => {

            Logger.logError({
              errorType: "AsynchronousValidationFailedError",
              title: "Asynchronous Validation Failed",
              description: `The asynchronous validation ${ validationRule.ID } has failed.`,
              occurrenceLocation: "inputtedValueValidation." +
                  "executeAsynchronousChecksIfAny(rawValue, currentValidationResult, asynchronousChecksCallback)",
              caughtError: error
            });

            asynchronousChecks[validationRule.ID] = {
              isPending: false,
              hasValidValueBeenConfirmed: false,
              hasInvalidValueBeenConfirmed: false,
              hasErrorOccurred: true,
              message: validationRule.messages.errorHasOccurred
            };

            asynchronousChecksCallback?.(
              new InputtedValueValidation.AsynchronousChecks.Status(asynchronousChecks),
              currentValidationResult
            );

          });

    }

  }

}


namespace InputtedValueValidation {

  export type Result = Readonly<
    {
      isValid: true;
    } |
    {
      isValid: false;
      errorsMessages: ReadonlyArray<string>;
    }
  >;


  /* ━━━ Rules ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  export interface Rule<NonEmptyValueType> {

    readonly mustFinishValidationIfValueIsInvalid: boolean;

    readonly check: (rawValue: NonEmptyValueType) => Rule.CheckingResult;

  }

  export namespace Rule {

    export type CheckingResult = Readonly<
      {
        isValid: true;
      } |
      {
        isValid: false;
        errorMessage: string;
      }
    >;

    export type ConstructorParameter = Readonly<{
      mustFinishValidationIfValueIsInvalid?: boolean;
    }>;

  }


  /* ┅┅┅ Asynchronous ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅ */
  export interface AsynchronousRule<NonEmptyValueType> {
    readonly ID: string;
    readonly messages: AsynchronousRule.Messages;
    readonly check: (rawValue: NonEmptyValueType) => Promise<AsynchronousRule.CheckingResult>;
  }

  export namespace AsynchronousRule {

    export type Messages = Readonly<{
      checkingInProgress: string;
      validValueHasBeenConfirmed: string;
      invalidValueHasBeenConfirmed: string;
      errorHasOccurred: string;
    }>;

    export type CheckingResult = Readonly<{
      isValid: boolean;
      errorMessage?: string;
    }>;

  }


  /* ━━━ Asynchronous Checking ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  export namespace AsynchronousCheck {

    export type Status = Readonly<{
      message: string;
      isPending: boolean;
      hasValidValueBeenConfirmed: boolean;
      hasInvalidValueBeenConfirmed: boolean;
      hasErrorOccurred: boolean;
    }>;

  }

  export type AsynchronousChecks = { [validationRuleName: string]: AsynchronousCheck.Status; };

  export namespace AsynchronousChecks {

    export type Callback = (status: Status, currentValidationResult: Result) => void;

    export class Status {

      public readonly checks: AsynchronousChecks;
      public readonly hasAtLeastOneCheckNotFinishedYet: boolean = false;
      public readonly hasAllChecksFinishedWithAnyOutcome: boolean = true;
      public readonly hasAtLeastOneCheckErrorOccurred: boolean = false;
      public readonly hasNoInvalidValuesBeenConfirmed: boolean = true;
      public readonly hasAtLeastOneInvalidValueBeenConfirmed: boolean = false;
      public readonly errorsMessages: Array<string> = [];

      public constructor(checks: AsynchronousChecks) {

        this.checks = checks;

        for (const checking of Object.values(checks)) {

          if (checking.isPending) {
            this.hasAtLeastOneCheckNotFinishedYet = true;
            this.hasAllChecksFinishedWithAnyOutcome = false;
          }

          if (checking.hasErrorOccurred) {
            this.hasAtLeastOneCheckErrorOccurred = true;
          }

          if (checking.hasInvalidValueBeenConfirmed) {
            this.hasNoInvalidValuesBeenConfirmed = false;
            this.hasAtLeastOneInvalidValueBeenConfirmed = true;
            this.errorsMessages.push(checking.message);
          }

        }

      }

    }

  }


  /* ━━━ Constructor Parameter ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  export type ConstructorCompoundParameter<NonEmptyValueType, EmptyValueType = NonEmptyValueType> =
      Readonly<{
        isValueOfSupportedType: SupportedValueChecker<NonEmptyValueType, EmptyValueType>;
        hasValueBeenOmitted: OmittedValueChecker<NonEmptyValueType, EmptyValueType>;
        isInputRequired: boolean | InputRequirementChecker;
        requiredInputIsMissingValidationErrorMessage?: string;
        staticRules?: ReadonlyArray<Rule<NonEmptyValueType>>;
        contextDependentRules?: ReadonlyArray<Rule<NonEmptyValueType>>;
        asynchronousRules?: ReadonlyArray<AsynchronousRule<NonEmptyValueType>>;
        asynchronousValidationsCallback?: AsynchronousChecks.Callback;
        localization?: Localization;
      }>;


  /* ━━━ Worktypes ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  export type SupportedValueChecker<NonEmptyValueType, EmptyValue = NonEmptyValueType> =
      (rawValue: unknown) => rawValue is NonEmptyValueType | EmptyValue;

  export type OmittedValueChecker<NonEmptyValueType, EmptyValue = NonEmptyValueType> =
      (supportedButPossiblyEmptyValue: NonEmptyValueType | EmptyValue) => supportedButPossiblyEmptyValue is EmptyValue;

  export type InputRequirementChecker = () => boolean;


  /* ━━━ Localization ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  export type Localization = Readonly<{
    requiredInputIsMissingValidationErrorMessage: string;
  }>;

}


export default InputtedValueValidation;

File: AdaptationsToFrameworks/Blazor/Package/GUI_Components/Controls/Validation/InputtedValueValidation.cs
namespace YamatoDaiwa.Frontend.GUI_Components.Controls.Validation;


public abstract class InputtedValueValidation
{

  public record Result
  {
    public required string[] ErrorsMessages { get; init; }
    public bool IsValid => this.ErrorsMessages.Length > 0;
  }
  
  
  /* ━━━ Rules ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  /* ┅┅┅ Interfaces ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅ */
  public interface IRule
  {
  
    public bool MustFinishValidationIfValueIsInvalid { get; init; }
  
    public IRule.CheckingResult Check(object rawValue);

    public record CheckingResult
    {
      public string? ErrorMessage { get; init; }
      public bool IsValid => this.ErrorMessage is not null;
    }
  
  }
  
  public interface IAsynchronousRule
  {
  
    public string ID { get; init; }
  
    
    /* ╍╍╍ Messages ╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍ */
    public record Messages
    {
      public required string CheckingInProgress { get; set; }
      public required string ValidValueHasBeenConfirmed { get; set; }
      public required string InvalidValueHasBeenConfirmed { get; set; }
      public required string ErrorHasOccurred { get; set; }
    }
    
    public IAsynchronousRule.Messages messages { get; init; }
    
  
    /* ╍╍╍ Checking ╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍ */
    public record CheckingResult
    {
      public string? ErrorMessage { get; init; }
      public bool IsValid => this.ErrorMessage is not null;
    }
    
    public Task<CheckingResult> Check(object rawValue);
    
  }
  
  
  /* ━━━ Asynchronous Checking ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public static class AsynchronousCheck
  {

    public record Status
    {
      public required string Message { get; init; }
      public required bool IsPending { get; init; }
      public required bool HasValidValueBeenConfirmed { get; init; } 
      public required bool HasInvalidValueBeenConfirmed { get; init; }
      public required bool HasErrorOccurred { get; init; }
    }
    
  }
  
  public static class AsynchronousChecks
  {

    public record Status
    {

      public readonly Dictionary<string, AsynchronousCheck.Status> Checks;
      public readonly bool HasAtLeastOneCheckNotFinishedYet = false;
      public readonly bool HasAllChecksFinishedWithAnyOutcome = true;
      public readonly bool HasAtLeastOneCheckErrorOccurred = false;
      public readonly bool HasNoInvalidValuesBeenConfirmed = true;
      public readonly bool HasAtLeastOneInvalidValueBeenConfirmed = false;
      public readonly string[] ErrorsMessages;

      public Status(Dictionary<string, AsynchronousCheck.Status> checks)
      {
        
        this.Checks = checks;

        List<string> errorsMessages = [];

        foreach ((string _, AsynchronousCheck.Status checking) in checks)
        {

          if (checking.IsPending)
          {
            this.HasAtLeastOneCheckNotFinishedYet = true;
            this.HasAllChecksFinishedWithAnyOutcome = false;
          }

          if (checking.HasErrorOccurred)
          {
            this.HasAtLeastOneCheckErrorOccurred = true;
          }

          if (checking.HasInvalidValueBeenConfirmed)
          {
            this.HasNoInvalidValuesBeenConfirmed = false;
            this.HasAtLeastOneInvalidValueBeenConfirmed = true;
            errorsMessages.Add(checking.Message);
          }

        }
        
        this.ErrorsMessages = errorsMessages.ToArray();
        
      }
    }
    
  }
  
  
  /* ━━━ Localization ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public interface ILocalization
  {
    public string RequiredInputIsMissingValidationErrorMessage { get; }
  }

  public static ILocalization Localization = new InputtedValueValidationEnglishLocalization();
  

  /* ━━━ Instance Fields ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public readonly Func<object?, bool> IsValueOfSupportedType;
  public readonly Func<object?, bool> HasValueBeenOmitted;
  public readonly Func<bool> IsInputRequired;

  protected readonly string RequiredInputIsMissingValidationErrorMessage;

  protected IRule[] StaticRules;
  protected IRule[] ContextDependentRules;
  protected IAsynchronousRule[] AsynchronousRules;
  
  protected Action<string>? asynchronousValidationFailureLogger;
  
  
  /* ━━━ Constructor ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  protected InputtedValueValidation(
    Func<object?, bool> isValueOfSupportedType,
    Func<object?, bool> hasValueBeenOmitted,
    bool? inputRequiredFlag = null,
    Func<bool>? isInputRequired = null,
    string? requiredInputIsMissingValidationErrorMessage = null,
    IRule[]? staticRules = null,
    IRule[]? contextDependentRules = null,
    IAsynchronousRule[]? asynchronousRules = null,
    Action<string>? asynchronousValidationFailureLogger = null
  )
  {

    this.IsValueOfSupportedType = isValueOfSupportedType;
    this.HasValueBeenOmitted = hasValueBeenOmitted;
    
    if (inputRequiredFlag is not null)
    {
      
      if (isInputRequired is not null)
      {
        throw new ArgumentException(
          "The \"inputRequiredFlag\" and \"isInputRequired\" parameters are incompatible. " +
          "Please specify one of them."
        );
      }
      
      this.IsInputRequired = () => inputRequiredFlag.Value;
      
    } else if (isInputRequired is not null)
    {
      this.IsInputRequired = isInputRequired;
    } else
    {
      throw new ArgumentException(
        "The value requirement not specified. " +
        "Specify \"inputRequiredFlag\" or \"isInputRequired\" but not both."
      );
    }
    
    this.RequiredInputIsMissingValidationErrorMessage =
        requiredInputIsMissingValidationErrorMessage ??
        InputtedValueValidation.Localization.RequiredInputIsMissingValidationErrorMessage;


    this.StaticRules = staticRules ?? [];
    this.ContextDependentRules = contextDependentRules ?? [];
    this.AsynchronousRules = asynchronousRules ?? [];

  }

  
  /* ━━━ Public Methods ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public Result Validate(
    object rawValue,
    bool mustPostponeAsynchronousValidation = false,
    Action<AsynchronousChecks.Status, Result>? asynchronousChecksCallback = null,
    IEnumerable<string>? messagesOfExternallyDetectedValidationErrors = null
  )
  {

    if (!this.IsValueOfSupportedType(rawValue))
    {
      throw new ArgumentException(
        $"The type \"{ rawValue.GetType() }\" of `rawValue` is incompatible with specified validators."
      );
    }
    
    
    bool isInputRequired = this.IsInputRequired();
    
    if (this.HasValueBeenOmitted(rawValue))
    {
      return new Result
      {
        ErrorsMessages = isInputRequired ? [ this.RequiredInputIsMissingValidationErrorMessage ] : [] 
      };
    }
    
    
    List<string> validationErrorsMessages = messagesOfExternallyDetectedValidationErrors?.ToList() ?? [];

    foreach (IRule staticValidationRule in this.StaticRules)
    {

      IRule.CheckingResult checkingCheckingResult = staticValidationRule.Check(rawValue);

      if (checkingCheckingResult.ErrorMessage is not null)
      {

        validationErrorsMessages.Add(checkingCheckingResult.ErrorMessage);

        if (staticValidationRule.MustFinishValidationIfValueIsInvalid)
        {
          break;
        }

      }
      
    }

    if (validationErrorsMessages.Count > 0)
    {
      return new Result
      {
        ErrorsMessages = validationErrorsMessages.ToArray()
      };
    }
    
    
    foreach (IRule contextDependentValidationRule in this.ContextDependentRules)
    {

      IRule.CheckingResult checkingCheckingResult = contextDependentValidationRule.Check(rawValue);

      if (checkingCheckingResult.ErrorMessage is not null)
      {

        validationErrorsMessages.Add(checkingCheckingResult.ErrorMessage);

        if (contextDependentValidationRule.MustFinishValidationIfValueIsInvalid)
        {
          break;
        }

      }
      
    }
    
    if (validationErrorsMessages.Count > 0)
    {
      return new Result
      {
        ErrorsMessages = validationErrorsMessages.ToArray()
      };
    }


    Result validationResult = new() { ErrorsMessages = [] };
    
    if (!mustPostponeAsynchronousValidation)
    {
      this.executeAsynchronousChecksIfAny(rawValue, validationResult, asynchronousChecksCallback); 
    }
    
    return validationResult;

  }

  public void executeAsynchronousChecksIfAny(
    object rawValue, 
    Result currentValidationResult, 
    Action<AsynchronousChecks.Status, Result>? asynchronousChecksCallback
  )
  {

    if (this.AsynchronousRules.Length == 0)
    {
      return;
    }


    Dictionary<string, AsynchronousCheck.Status> asynchronousChecks = this.AsynchronousRules.
        ToDictionary(
          asynchronousRule => asynchronousRule.ID, 
          asynchronousRule => new AsynchronousCheck.Status
          {
            Message = asynchronousRule.messages.CheckingInProgress,
            IsPending = true,
            HasInvalidValueBeenConfirmed = false,
            HasValidValueBeenConfirmed = false,
            HasErrorOccurred = false
          }
        );

    asynchronousChecksCallback?.Invoke(new AsynchronousChecks.Status(asynchronousChecks), currentValidationResult);

    IEnumerable<Task> validationTasks = this.AsynchronousRules.Select(
      async validationRule =>
      {

        IAsynchronousRule.CheckingResult checkingResult;
        
        try
        {

          checkingResult = await validationRule.Check(rawValue);

        }
        catch (Exception exception)
        {

          this.asynchronousValidationFailureLogger?.Invoke(
            $"The asynchronous validation { validationRule.ID } has failed.\n" +
            exception.Message
          );
          
          asynchronousChecks[validationRule.ID] = new AsynchronousCheck.Status
          {
            IsPending = false,
            HasValidValueBeenConfirmed = false,
            HasInvalidValueBeenConfirmed = false,
            HasErrorOccurred = true,
            Message = validationRule.messages.ErrorHasOccurred
          };
          
          asynchronousChecksCallback?.Invoke(new AsynchronousChecks.Status(asynchronousChecks), currentValidationResult);
          
          return;
          
        }
        
        
        asynchronousChecks[validationRule.ID] = new AsynchronousCheck.Status
        {
          IsPending = false,
          HasValidValueBeenConfirmed = checkingResult.IsValid,
          HasInvalidValueBeenConfirmed = !checkingResult.IsValid,
          HasErrorOccurred = false,
          Message = checkingResult.IsValid ? 
              validationRule.messages.ValidValueHasBeenConfirmed :
              checkingResult.ErrorMessage ?? validationRule.messages.InvalidValueHasBeenConfirmed 
        };
        
        AsynchronousChecks.Status asynchronousChecksStatus = new(asynchronousChecks);

        string[] errorsMessages = currentValidationResult.ErrorsMessages.
            Concat(asynchronousChecksStatus.ErrorsMessages).
            ToArray();

        asynchronousChecksCallback?.Invoke(asynchronousChecksStatus, new Result { ErrorsMessages = errorsMessages });

      }
    );

    Task.WhenAll(validationTasks);

  }
  
}